controllers/appsource_controller.go

```diff
@@ -0,0 +1,145 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"errors"
+	"regexp"
+
+	clusterv1 "github.com/argoproj-labs/argocd-app-source/api/v1"
+
+	//?Are these imports correct? They seem to be throwing an error.
+	appclientset "github.com/argoproj/argo-cd/pkg/client/clientset/versioned"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// AppSourceReconciler reconciles a AppSource object
+type AppSourceReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+
+	//? I feel like I'm not using the reconciler to its fullest potential.
+	//? The reconcile logic will always need the ArgoCD client as well as
+	//? the Kubernetes client, maybe I can add them to the reconciler type?
+	//? But then how will the ArgoCD client by dynamically initialized based
+	//? on the address provided in the AppSource ConfigMap?
+	ArgoAppClientset appclientset.Interface
+	KubeClientset    kubernetes.Interface
+}
+
+const (
+	appsource_cm_name = "argocd-sourc-cm"
+	argocd_namespace  = "argocd"
+)
+
+//+kubebuilder:rbac:groups=cluster.my.domain,resources=appsources,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=cluster.my.domain,resources=appsources/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=cluster.my.domain,resources=appsources/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the AppSource object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *AppSourceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = log.FromContext(ctx)
+
+	pattern_matches_namespace, err := r.ValidateNamespacePattern(ctx, req)
+	if err != nil {
+		panic(err)
+	}
+
+	if pattern_matches_namespace {
+		err = r.ValidateProject(ctx, req)
+		if err != nil {
+			panic(err)
+		}
+		err = r.ValidateApplication(ctx, req)
+		if err != nil {
+			panic(err)
+		}
+	} else {
+		//Name does not match namespace regex pattern.
+		panic(errors.New("Namespace does not match AppSource Project Pattern."))
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *AppSourceReconciler) ValidateApplication(ctx context.Context, req ctrl.Request) (err error) {
```

Comment:
Methods `ValidateApplication` and others are used only inside of `AppSourceReconciler` and don't have to be public. Please rename to `validateApplication` etc

Reply:
Thanks Alex I wasn't aware of the capital letters exporting functions  — I'll make these changes